### PR TITLE
ISPN-14946/47/51/52/53 RESP MEMORY USAGE|STATS, TYPE, TTL, PTTL

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-14946 MEMORY USAGE
https://issues.redhat.com/browse/ISPN-14947 MEMORY STATS
https://issues.redhat.com/browse/ISPN-14951 TYPE
https://issues.redhat.com/browse/ISPN-14952 TTL
https://issues.redhat.com/browse/ISPN-14953 PTTL

This allows RedisInsight to work.

### CODE REVIEW

**Comment by tristantarrant** [comment]
Failures are unrelated

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/MEMORY.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+public class MEMORY extends RespCommand implements Resp3Command {
+   public MEMORY() {
+      super(-2, 0, 0, 0);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx,
+                                                      List<byte[]> arguments) {
+      handler.checkPermission(AuthorizationPermission.ADMIN);
+      String subcommand = new String(arguments.get(0), StandardCharsets.UTF_8).toUpperCase();
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/connection/MEMORY.java:34]: I believe all commands have to be US_ASCII afaik.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/MEMORY.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+            addStat(sb, "fragmentation.bytes", 0); //53
+            ByteBufferUtils.stringToByteBuf(sb, handler.allocator());
+            break;
+         case "USAGE":
+            if (arguments.size() < 2) {
+               RespErrorUtil.wrongArgumentCount(this, handler.allocator());
+               return handler.myStage();
+            } else {
+               byte[] key = arguments.get(1);
+               return handler.stageToReturn(handler.cache().getAsync(key).thenApply(v -> v == null ? null : (long) (key.length + v.length + 14) / 8 * 8), ctx, Consumers.LONG_BICONSUMER);
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/connection/MEMORY.java:74]: I would separate this to a different line, kinda hard to read as is.

File: server/resp/src/main/java/org/infinispan/server/resp/commands/generic/PTTL.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.commands.Resp3Command;
+
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * PTTL Resp Command
+ * h<a href="ttps://redis.io/commands/pttl/">pttl</a>
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/generic/PTTL.java:16]:
```suggestion
 * <a href="https://redis.io/commands/pttl">pttl</a>
```

File: server/resp/src/main/java/org/infinispan/server/resp/commands/generic/TTL.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.commands.Resp3Command;
+
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * TTL Resp Command
+ * h<a href="ttps://redis.io/commands/ttl/">ttl</a>
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/generic/TTL.java:16]:
```suggestion
 * <a href="https://redis.io/commands/ttl">ttl</a>
```

File: server/resp/src/main/java/org/infinispan/server/resp/commands/generic/TYPE.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
+                                                      ChannelHandlerContext ctx,
+                                                      List<byte[]> arguments) {
+      byte[] keyBytes = arguments.get(0);
+
+      return handler.stageToReturn(handler.cache().getCacheEntryAsync(keyBytes).thenApply(e -> {
+         if (e == null) {
+            return RespTypes.none.name().getBytes(StandardCharsets.US_ASCII);
```
Comment (wburns) [inline server/resp/src/main/java/org/infinispan/server/resp/commands/generic/TYPE.java:34]: Do we want to cache these byte[] objects?

**Review by wburns — COMMENTED** [review]
LGTM, just some very minor stuff

File: server/resp/src/main/java/org/infinispan/server/resp/commands/generic/TYPE.java (hunk 5 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/generic/TYPE.java:34]: I don't think this is used in performance-critical apps

File: server/resp/src/main/java/org/infinispan/server/resp/commands/generic/TTL.java (hunk 4 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/generic/TTL.java]: Done

File: server/resp/src/main/java/org/infinispan/server/resp/commands/generic/PTTL.java (hunk 3 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/generic/PTTL.java]: Done

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/MEMORY.java (hunk 2 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/connection/MEMORY.java]: Done

File: server/resp/src/main/java/org/infinispan/server/resp/commands/connection/MEMORY.java (hunk 1 above)
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/commands/connection/MEMORY.java]: Done

**Review by wburns — COMMENTED** [review]
LGTM, waiting on CI.

**Comment by tristantarrant** [comment]
CI was done 2 days ago :)

**Comment by wburns** [comment]
Integrated into main, thanks @tristantarrant !

**Pull request merged**